### PR TITLE
Faster `Layout::array`

### PR DESCRIPTION
**Pull request by nnethercote**
`Layout::array` is called (indirectly) by `Vec::push()`, which is typically instantiated many times, and so making it smaller can help with compile times because less LLVM IR is generated.

r? @ghost

### CODE REVIEW

**Comment by nnethercote** [comment]
Instructions mostly show a win, with a few small regressions. Cycles and wall-times show no regressions but fewer wins.

We've seen in the past that this `Vec::push` path is really finicky and LLVM's choices of how to inline things can have surprising results (e.g. https://github.com/rust-lang/rust/pull/72189). So I am happy this result. It's clear that `Layout::array` is shorter and better, which in general can only result in faster compile times and runtimes, even if in some cases random fluctuations mask the effect.

r? @Amanieu

**Comment by nnethercote** [comment]
BTW I confirmed with `cargo llvm-lines` that this results in fewer lines of LLVM IR.

**Review by dtolnay — APPROVED** [review]
Nice!

**Comment by dtolnay** [comment]
```console
fatal error LNK1201: error writing to program database 'D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\tmp\cit\t1300\foo\target\release\deps\test.pdb'; check for insufficient disk space, invalid path, or insufficient privilege
```

@bors retry

File: library/core/src/alloc/layout.rs [hunk 1]
```diff
@@ -403,9 +405,17 @@ impl Layout {
...
+        let array_size = mem::size_of::<T>().checked_mul(n).ok_or(LayoutError)?;
+
+        // SAFETY:
+        // - Size: `array_size` cannot be too big because `size_of::<T>()` must
+        //   be a multiple of `align_of::<T>()`. Therefore, `array_size`
+        //   rounded up to the nearest multiple of `align_of::<T>()` is just
+        //   `array_size`. And `array_size` cannot be too big because it was
+        //   just checked by the `checked_mul()`.
+        // - Alignment: `align_of::<T>()` will always give an acceptable
+        //   (non-zero, power of two) alignment.
```
Comment (the8472) [inline library/core/src/alloc/layout.rs:417]: Is this true? I vaguely recall some conversation that this only applies to `repr(Rust)` types buy that some weird extern types could have non-power of two alignments. But I can't find a source on that, so I may be misremembering.

Comment (dtolnay) [reply library/core/src/alloc/layout.rs:417]: This is fine according to https://doc.rust-lang.org/1.56.0/reference/type-layout.html#size-and-alignment, which doesn't say it's specific to repr(Rust).

> Alignment is measured in bytes, and must be at least 1, and always a power of 2. The alignment of a value can be checked with the `align_of_val` function.

Comment (the8472) [reply library/core/src/alloc/layout.rs:417]: Nevermind, Layout itself requires it and so do C and C++, so I must have been thinking of something else.

**Pull request merged**